src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,515 @@ func (s *commitLogSource) ReadData(
+				// another copy of the IDs and tags.
+				id.Finalize()
+				id = existing.ID
+				tags = existing.Tags
+				shouldDecodeTags = false
+			}
+		}
+
+		if shouldDecodeTags {
+			// Only spend cycles decoding the tags if we've never seen them before.
+			if tagsIter.Remaining() > 0 {
+				tags, err = convert.TagsFromTagsIter(id, tagsIter, idPool)
+				if err != nil {
+					return shardResult, fmt.Errorf("unable to decode tags: %v", err)
+				}
+			}
+			tagsIter.Close()
+		}
+
+		if shardResult == nil {
+			// Delay initialization so we can estimate size.
+			shardResult = result.NewShardResult(reader.Entries(), s.opts.ResultOptions())
+		}
+
+		// Mark the ID and Tags as no finalize to enable no-copy optimization later
+		// in the bootstrap process (when they're being loaded into the shard). Also,
+		// technically we'll be calling NoFinalize() repeatedly on the same IDs for
+		// different blocks since we're reusing them, but thats ok as it an idempotent
+		// operation and there is no concurrency here.
+		id.NoFinalize()
+		tags.NoFinalize()
+		shardResult.AddBlock(id, tags, dbBlock)
+	}
+
+	return shardResult, nil
+}
+
+func (s *commitLogSource) newReadCommitLogPredBasedOnAvailableSnapshotFiles(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	snapshotFilesByShard map[uint32]fs.FileSetFilesSlice,
+) (
+	func(fileName string, fileStart time.Time, fileBlockSize time.Duration) bool,
+	map[xtime.UnixNano]map[uint32]fs.FileSetFile,
+	error,
+) {
+	blockSize := ns.Options().RetentionOptions().BlockSize()
+
+	// At this point we've bootstrapped all the snapshot files that we can, and we need to
+	// decide which commit logs to read. In order to do that, we'll need to figure out the
+	// minimum most recent snapshot time for each block, then we can use that information to
+	// decide how much of the commit log we need to read for each block that we're bootstrapping.
+	// To start, for each block that we're bootstrapping, we need to figure out the most recent
+	// snapshot that was taken for each shard. I.E we want to create a datastructure that looks
+	// like this:
+	// 		map[blockStart]map[shard]mostRecentSnapshotTime
+	mostRecentCompleteSnapshotByBlockShard := s.mostRecentCompleteSnapshotByBlockShard(
+		shardsTimeRanges, blockSize, snapshotFilesByShard, s.opts.CommitLogOptions().FilesystemOptions())
+	for block, mostRecentByShard := range mostRecentCompleteSnapshotByBlockShard {
+		for shard, mostRecent := range mostRecentByShard {
+
+			if mostRecent.CachedSnapshotTime.IsZero() {
+				// Should never happen.
+				return nil, nil, fmt.Errorf(
+					"%s shard: %d and block: %d had zero value for most recent snapshot time",
+					instrument.InvariantViolatedMetricName, shard, block.ToTime().Unix())
+			}
+
+			s.log.Infof(
+				"most recent snapshot for block: %d and shard: %d is %d",
+				block.ToTime().Unix(), shard, mostRecent.CachedSnapshotTime.Unix())
```

Comment:
Perhaps best to log these dates as string like we do elsewhere?

Reply:
done